pkg/datamover/repository_server.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datamover
+
+import (
+	"context"
+	"github.com/pkg/errors"
+
+	"github.com/kanisterio/kanister/pkg/kopia/repository"
+	"github.com/kanisterio/kanister/pkg/kopia/snapshot"
+	"github.com/kanisterio/kanister/pkg/param"
+)
+
+// Check that RepositoryServer implements DataMover interface
+var _ DataMover = (*RepositoryServer)(nil)
+
+type RepositoryServer struct {
+	outputName       string
+	repositoryServer *param.RepositoryServer
+	snapJSON         string
+}
+
+func (rs *RepositoryServer) Pull(ctx context.Context, sourcePath, destinationPath string) error {
+	kopiaSnap, password, err := rs.kopiaSnapshotAndRepositoryServerUserPassword(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationPull(ctx, kopiaSnap.ID, destinationPath, sourcePath, password)
+}
+
+func (rs *RepositoryServer) Push(ctx context.Context, sourcePath, destinationPath string) error {
+	password, err := rs.connectToKopiaRepositoryServer(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationPush(ctx, destinationPath, rs.outputName, sourcePath, password)
+}
+
+func (rs *RepositoryServer) Delete(ctx context.Context, destinationPath string) error {
+	kopiaSnap, password, err := rs.kopiaSnapshotAndRepositoryServerUserPassword(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationDelete(ctx, kopiaSnap.ID, destinationPath, password)
+}
+
+func (rs *RepositoryServer) connectToKopiaRepositoryServer(ctx context.Context) (string, error) {
+	hostname, userPassphrase, certData, err := secretsFromRepositoryServerCR(rs.repositoryServer)
+	if err != nil {
+		return "", errors.Wrap(err, "Error Retrieving Connection Data from Repository Server")
+	}
+	return userPassphrase, repository.ConnectToAPIServer(
+		ctx,
+		certData,
+		userPassphrase,
+		hostname,
+		rs.repositoryServer.Address,
+		rs.repositoryServer.Username,
+		rs.repositoryServer.ContentCacheMB,
+		rs.repositoryServer.MetadataCacheMB,
+	)
+}
+
+func (rs *RepositoryServer) kopiaSnapshotAndRepositoryServerUserPassword(ctx context.Context) (*snapshot.SnapshotInfo, string, error) {
```

Comment:
This function is doing multiple unrelated things. Can we split this into logical utils?

Reply:
Sure !